Camera class added and fiducial marker detections

chris24sahadeo opened the pull request:
Milestone 1a v1

chris24sahadeo committed:
Testing package

chris24sahadeo committed:
Added camera class.

End of milestone 1 draft 1

chris24sahadeo committed:
ROSCpp Code Format

sarika93 commented on rviz_simulator/launch/simulate.launch:
```diff
@@ -0,0 +1,8 @@
+  <rosparam command="load" file="$(find rviz_simulator)/config/initialize_simulator.yaml" />
+  <rosparam command="load" file="$(find rviz_simulator)/config/realsense_camera_intrinsics.yaml" />
+  <node pkg="rviz" type="rviz" name="rviz"/>
+  <node pkg="rqt_console" type="rqt_console" name="rqt_console"/>
```
rqt is a debugging tool, so this launching this every time you want to use this application isn't necessary. Remove this from the launch file. The same applies to rqt_logger_level.

chris24sahadeo replied:
I have made a new file "simulate_debug.launch" because I use rqt often during testing.

i'll make a separate file named "simulate.launch" to use when not debugging

sarika93 replied:
Instead of doing that, add a command line arg (e.g. debug_mode), when if debug_mode is true, then you launch these, otherwise, do not. 
See [here](http://wiki.ros.org/roslaunch/XML) for if conditions in launch files (Section 3) and [here](http://wiki.ros.org/roslaunch/XML/arg) for args.

chris24sahadeo replied:
Done.

sarika93 commented on rviz_simulator/include/rviz_simulator/target.h:
```diff
@@ -42,36 +41,46 @@ namespace rviz_simulator
+  /// @param marker_color_RGBA                Marker colour (red, green blue) and opacity setting
+  /// @param marker_scale                     The length of a side of the interactive marker
+  /// @param g_interactive_marker_server      Shared pointer for interactive marker server
+  /// @oaram interaction_mode                 Specifies how the marker will react to events (3D MOVEMENT or BUTTON
```
Typo.

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/src/target.cpp:
```diff
@@ -84,15 +85,16 @@ void Target::make6DofMarker(const geometry_msgs::Point& position)
 
   this->interactive_marker_.name = this->marker_name_;
-  this->interactive_marker_.description = "6DOF_MOVE_ROTATE_3D";
+  this->interactive_marker_.description = this->marker_name_;
```
The description being the same as the name isn't useful. If this field can remain empty, then you can simply leave it empty.

chris24sahadeo replied:
The description value is the label that shows up under each marker in rviz.

The name is the value we use to get interactive markers from the interactive marker server.

So I think that its okay to keep them as the same value. They both serve different but useful purposes.

sarika93 replied:
Ah, ok I didn't realize this would be the visible name in RViz (I thought it was name). This is fine then.

sarika93 commented on rviz_simulator/src/target.cpp:
```diff
@@ -164,10 +166,11 @@ void Target::processFeedback(const visualization_msgs::InteractiveMarkerFeedback
+                                                  boost::bind(&rviz_simulator::Target::targetFeedback, this, _1));
+  ROS_INFO_STREAM("Target " << this->marker_name_ << " added.\n");
 }
 }
```
Add newline at the end of all files.

chris24sahadeo replied:
Fixed in all files.

sarika93 commented on rviz_simulator/include/rviz_simulator/camera.h:
```diff
@@ -0,0 +1,183 @@
+
+/// Camera distortions. Nomenclature follows from the OpenCV docs
+/// https://docs.opencv.org/2.4/modules/calib3d/doc/camera_calibration_and_3d_reconstruction.html
+struct CameraDistortions
```
It looks like this can be one struct since every camera will always have one CameraIntrinsics and one CameraDistortions object. Merge these into one struct called CameraProperties. Move the camera_matrix_ into CameraProperties from Camera as a private variable, and include a public function (e,g. public void populate_cam_matrix()) to populate camera_matrix_. 

The camera matrix has a fixed format, so having fx, fy etc is already enough to automatically create the matrix.

chris24sahadeo replied:
Fixed.

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/include/rviz_simulator/camera.h:
```diff
@@ -0,0 +1,183 @@
+  /// Camera properties
+  CameraIntrinsics camera_intrinsics_;
+  CameraDistortions camera_distortions_;
+  Eigen::MatrixXd camera_intrisics_matrix_;
```
Typo but you can rename this to camera_matrix; this implicitly indicates intrinsics.

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/include/rviz_simulator/camera.h:
```diff
@@ -0,0 +1,183 @@
+  CameraDistortions camera_distortions_;
+  Eigen::MatrixXd camera_intrisics_matrix_;
+  double length_of_target_;         /// The real-world length of a target
+  int detections_file_number_;      /// The number of pictures taken during the running of the node
```
Rename to num_pictures_ or similar so that it is more obvious what this variable represents.

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/include/rviz_simulator/camera.h:
```diff
@@ -0,0 +1,183 @@
+  CameraIntrinsics camera_intrinsics_;
+  CameraDistortions camera_distortions_;
+  Eigen::MatrixXd camera_intrisics_matrix_;
+  double length_of_target_;         /// The real-world length of a target
```
Rename to target_length_ since this is shorter.

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/src/camera.cpp:
```diff
@@ -0,0 +1,361 @@
+  this->addShutterReleaseButton();
+
+  // create ouput directory for yaml files
+  int detections_folder_number_ = int(ros::Time::now().toSec());
```
Local variables should not have underscores at the end. Those are for member variables.

chris24sahadeo replied:
Fixed

sarika93 commented on rviz_simulator/src/camera.cpp:
```diff
@@ -0,0 +1,361 @@
+  // create ouput directory for yaml files
+  int detections_folder_number_ = int(ros::Time::now().toSec());
+  this->detections_file_number_ = 0;
+  this->output_folder_name_ = "detections_" + std::to_string(detections_folder_number_);
```
output_folder_name_ doesn't need to be a member variable as it's only used locally in this function.

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/src/camera.cpp:
```diff
@@ -0,0 +1,361 @@
+  this->g_interactive_marker_server_->applyChanges();
+}
+
+void Camera::addShutterReleaseButton()
```
Bit of overkill to have a function that is one line that gets called once in the constructor. Get rid of this function and just add a comment saying "Adding a shutter release button".

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/src/camera.cpp:
```diff
@@ -0,0 +1,361 @@
+  this->interactive_marker_.controls[0].interaction_mode = visualization_msgs::InteractiveMarkerControl::BUTTON;
+}
+
+Eigen::Affine3d Camera::getTransformBetweenInteractiveMarkers(visualization_msgs::InteractiveMarker& reference,
```
Rename to calculateReference_T_Target or calculateMarker1_T_Marker2. Get is generally used to retrieve an already set variable. TransformBetweenInteractiveMarkers is not clear; Marker1_T_Marker2 clearly shows that the transform is used to transform something from Marker2's frame into the Marker1 frame.

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/src/camera.cpp:
```diff
@@ -0,0 +1,361 @@
+Eigen::Affine3d Camera::getTransformBetweenInteractiveMarkers(visualization_msgs::InteractiveMarker& reference,
+                                                              visualization_msgs::InteractiveMarker& target)
+{
+  Eigen::Affine3d reference_pose, target_pose;
```
Use transform notation for these variables so it is clearer what you are hoping to achieve:
e.g. Eigen::Affine3d origin_T_reference, origin_T_pose. You can replace origin with ROSWorld if you think that is clearer.

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/src/camera.cpp:
```diff
@@ -0,0 +1,361 @@
+  return reference_pose.inverse() * target_pose;
+}
+
+void transformToYAMLSequence(Eigen::Affine3d& transform, YAML::Emitter& out)
```
This function name doesn't need to be so long. Affine3dToYAML or Affine3dToYAMLSeq is fine. Don't use "Transform" as this function doesn't need to be limited to transforms.

The output format of this function is bad since it doesn't actually show that this is a matrix. Format this to be something like the following:
[ [a11, a12, a13, a14], [a21, a22, a23, a24], [a31, a32, a33, a34], [a41, a42, a43, a44] ]

Alternatively, rename to transformToYAMLSeq, convert the matrix to position and rotation (in either Rodriguez or Quaternion format) and output in the format:

position: [x, y, z]
rotation: [x, y, z, w]

sarika93 commented on rviz_simulator/src/camera.cpp:
```diff
@@ -0,0 +1,361 @@
+  }                     // end processing all camera_T_target transforms
+  out << YAML::EndSeq;  // level 1
+
+  // publishing camera pose
```
Publishing is a bad word to use here since ROS uses ROS publishers and that is not what you are doing here.

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/src/target.cpp:
```diff
@@ -164,10 +166,11 @@ void Target::processFeedback(const visualization_msgs::InteractiveMarkerFeedback
   this->g_interactive_marker_server_->setCallback(this->interactive_marker_.name,
-                                                  boost::bind(&rviz_simulator::Target::processFeedback, this, _1));
+                                                  boost::bind(&rviz_simulator::Target::targetFeedback, this, _1));
+  ROS_INFO_STREAM("Target " << this->marker_name_ << " added.\n");
```
Make this output more explicit to state "....added to server. \n"

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/src/camera.cpp:
```diff
@@ -0,0 +1,361 @@
+  Eigen::Affine3d world_T_camera = this->getTransformBetweenInteractiveMarkers(world_marker, camera_marker);
+
+  out << YAML::Key << "camera" << YAML::Value << YAML::BeginMap;  // level 0
+  out << YAML::Key << "pose" << YAML::Value << YAML::BeginSeq;
```
rename pose to "world_T_camera". Pose just isn't clear.

sarika93 commented on rviz_simulator/src/camera.cpp:
```diff
@@ -0,0 +1,361 @@
+        // getting world_T_target
+        // outputting world_T_target to YAML row_wise
+        Eigen::Affine3d world_T_target = this->getTransformBetweenInteractiveMarkers(world_marker, target_marker);
+        out << YAML::Key << "pose" << YAML::Value << YAML::BeginSeq;
```
Rename "pose" to world_T_tag[i]

sarika93 commented on rviz_simulator/src/camera.cpp:
```diff
@@ -0,0 +1,361 @@
+  }
+}
+
+std::vector<Eigen::Vector2d> Camera::processCorners(Eigen::Affine3d camera_extrinsics)
```
camera_extrinsics is a bad name as it can refer to cam to tag or world and does not indicate direction. Use cam_T_tag or similar instead.

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/src/camera.cpp:
```diff
@@ -0,0 +1,361 @@
+
+void Camera::takePicture()
+{
+  YAML::Emitter out;
```
YAML format should look more like David's. You can also check the Project Description document for the format.

sarika93 commented on rviz_simulator/launch/simulate.launch:
```diff
@@ -0,0 +1,8 @@
+<launch>
+  <rosparam command="load" file="$(find rviz_simulator)/config/initialize_simulator.yaml" />
+  <rosparam command="load" file="$(find rviz_simulator)/config/realsense_camera_intrinsics.yaml" />
+  <node pkg="rviz" type="rviz" name="rviz"/>
```
Launch rviz using a config file. This is done as follows:
node type="rviz" name="rviz_sim" pkg="rviz" args="-d $(find package_name)/rviz/config_file.rviz"

This way you can have a preconfigured rviz setup and not need to add topics, frames etc every time.

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/src/camera.cpp:
```diff
@@ -0,0 +1,361 @@
+
+std::vector<Eigen::Vector2d> Camera::processCorners(Eigen::Affine3d camera_extrinsics)
+{
+  Eigen::Vector4d c1, c2, c3, c4;
```
Add in a comment saying that c1 is the bottom left corner of a tag and that the corners are assigned in anti-clockwise order.

chris24sahadeo replied:
Comment added.
This is subject to change according to David's research on the AprilTag library.

sarika93 commented on rviz_simulator/src/camera.cpp:
```diff
@@ -0,0 +1,361 @@
+  // without distortion rectification
+  // Eigen::Vector3d pixel_coordinates3d = this->camera_intrisics_matrix_*camera_extrinsics.matrix()*point;
+
+  // with distortion rectification
```
Provide a link to your source

chris24sahadeo replied:
Link to the source of what?

sarika93 replied:
Source of these formulae (I assume the opencv documentation?)

sarika93 commented on rviz_simulator/src/camera.cpp:
```diff
@@ -0,0 +1,361 @@
+  // Eigen::Vector3d pixel_coordinates3d = this->camera_intrisics_matrix_*camera_extrinsics.matrix()*point;
+
+  // with distortion rectification
+  Eigen::Vector4d pixel_coordinates3d = camera_extrinsics * point;
```
pixel_coordinates3d is not a good name. This is actually the point transformed into the camera frame. Use cam_point or similar instead. In that same vein, rename point to tag_point for clarity.

sarika93 replied:
Based on our discussion today rename to point_in_camera and point_in_tag instead of what was previously suggested.

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/src/camera.cpp:
```diff
@@ -0,0 +1,361 @@
+  double u = camera_intrinsics_.fx * x_double_prime + camera_intrinsics_.cx;
+  double v = camera_intrinsics_.fy * y_double_prime + camera_intrinsics_.cy;
+
+  Eigen::Vector2d pixelCoordinates2d;
```
You can just name this pixelCoordinates. 2D is understood.

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/src/camera.cpp:
```diff
@@ -0,0 +1,361 @@
+  return corners2d;
+}
+
+Eigen::Vector2d Camera::toPixelCoordinates(Eigen::Affine3d camera_extrinsics, Eigen::Vector4d point)
```
This is a bad function name. What are you converting to pixel coordinates? Rename to calculatePixelCoords

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/src/camera.cpp:
```diff
@@ -0,0 +1,361 @@
+  return pixelCoordinates2d;
+}
+
+bool Camera::isValidCorner(int x, int y)
```
Use Vector2d as the input type since data is already in that format where this is being used. Consider renaming to isWithinImage since other factors can invalidate a corner.

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/src/camera.cpp:
```diff
@@ -0,0 +1,361 @@
+  // TODO
+}
+
+bool Camera::isInRange(Eigen::Affine3d transform)
```
We need to discuss this as this behaviour is not realistic. This function isn't a good approach for determining if something is too far away to be detected.

sarika93 reviewed (changes requested):
Not done reviewing this, but there is definitely enough here to tackle while I finish reviewing the rest of it. 
The YAML formatting needs a big refactor, as well as determining if something is too far away.

chris24sahadeo committed:
Commit before making "Big Discussion" changes

TODO:
- Function to detect if target is too far away

- Change code to confirm to new YAML output requirements

- Document desired YAML output with David

chris24sahadeo committed:
Midway commit.

Still working on milestone 1 Big Discussion Fixes.
Committing just to save progress.

chris24sahadeo committed:
FIXME:

The camera detects more targets on the right side then the left side

chris24sahadeo committed:
REVIEW ME

Code now generates the new YAML file formats,
as specified in the google doc.

chris24sahadeo committed:
Changed target orientations.

sarika93 commented on rviz_simulator/include/rviz_simulator/camera.h:
```diff
@@ -0,0 +1,231 @@
+  /// camera image extremes
+  int image_height;
+  int image_width;
+  double min_distance_between_target_corners;  // in pixels
```
This is a very long name. Consider shortening e.g. min_corner_dist.

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/include/rviz_simulator/camera.h:
```diff
@@ -0,0 +1,231 @@
+  int image_height;
+  int image_width;
+  double min_distance_between_target_corners;  // in pixels
+  double max_distance_between_camera_and_target;
```
Shorten name e.g. cam_target_min_dist

chris24sahadeo replied:
Fixed

sarika93 commented on rviz_simulator/include/rviz_simulator/camera.h:
```diff
@@ -0,0 +1,231 @@
+  int image_width;
+  double min_distance_between_target_corners;  // in pixels
+  double max_distance_between_camera_and_target;
+  double min_distance_between_camera_and_target;
```
Shorten name

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/include/rviz_simulator/camera.h:
```diff
@@ -0,0 +1,231 @@
+  /// Calculates the pixel coordinates of the four corners of a target.
+  /// @param camera_extrinsics    The transform camera_T_target.
+  /// @return                     Vector of corner (x, y) pixel coordinates
+  std::vector<Eigen::Vector2d> processCorners(Eigen::Affine3d camera_extrinsics);
```
Instead of explaining that camera_extrinsics is camera_T_target in a comment, rename it to camera_T_target. It's much easier for someone to follow that way.

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/launch/simulate.launch:
```diff
@@ -0,0 +1,19 @@
+<launch>
+  <!-- loading values into the ros parameter server -->
```
Move all of the arg definitions (e.g.  arg name="debug_mode" default="false") here so it easier for a user to check the file and see what args this file can use.

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/config/realsense_camera_intrinsics.yaml:
```diff
@@ -0,0 +1,22 @@
+image_width: 1280
+image_height: 720
+camera_name: "Realsense Camera (Camera)"
```
The "(Camera)" is unnecessary since it's already called "Realsense Camera"

chris24sahadeo replied:
Fixed.

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/README.md:
```diff
@@ -0,0 +1,45 @@
+To enter `debug_mode`, run the launch file with the command line arg 
+> `debug_mode:=true`
+
+By default, the `realsense_camera_intrinsics.yaml` file is loaded. To specify another camera intrinsics file, run the launch file with the command line arg
```
photoneo by default.

chris24sahadeo replied:
Fixed

sarika93 commented on rviz_simulator/src/simulate.cpp:
```diff
@@ -29,13 +29,125 @@
+std_msgs::ColorRGBA loadColor(const ros::NodeHandle& n, const std::string& color_name)
+{
+  std::vector<double> color_vector;
+  n.getParam(color_name, color_vector);
```
Consider using param instead of getParam as this allows you to use a default if the user does not set this param.

chris24sahadeo replied:
Deleted the `loadColor` function.
Hard coding colors instead.

sarika93 commented on rviz_simulator/src/simulate.cpp:
```diff
@@ -29,13 +29,125 @@
+
+/// Loads geometry_msgs::Point from ROS parameter server
+geometry_msgs::Point loadPoint(const ros::NodeHandle& n, const std::string& point_name)
+{
```
Use param and specify a default position if the user does not set this parameter.

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/src/simulate.cpp:
```diff
@@ -29,13 +29,125 @@
+/// Loads geometry_msgs::Quaternion from ROS parameter server
+geometry_msgs::Quaternion loadOrientation(const ros::NodeHandle& n, const std::string& quarternion_name)
+{
+  std::vector<double> quaternion_vector;
```
Use [param ](http://wiki.ros.org/roscpp_tutorials/Tutorials/Parameters)and specify a default orientation if the user does not set this parameter.

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/src/simulate.cpp:
```diff
@@ -29,13 +29,125 @@
+}
+
+/// Loads rviz_simulator::CameraProperties from ROS parameter server
+rviz_simulator::CameraProperties loadCameraProperties(const ros::NodeHandle& n)
```
If the user is missing any camera properties, you should output error messages saying which params are missing. Don't use any defaults in this function as the purpose of this exercise is to simulate some actual camera's behaviour. getParam allows you to check if a param was obtained or not.

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/src/simulate.cpp:
```diff
@@ -29,13 +29,125 @@
+  {
+    ROS_ERROR("Unknown camera distortion model specified!\n");
+  }
+  camera_properties.min_distance_between_target_corners = 30;
```
Hardcoding this in a function is a bad design. If this is always constant regardless of all other camera properties, assign in the CameraProperties class instead.

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/src/simulate.cpp:
```diff
@@ -45,26 +157,57 @@ int main(int argc, char** argv)
-  double marker_scale = 0.1;
+  // loading parameters
+  std::string world_frame_id;
+  n.getParam("/world_frame_id", world_frame_id);
```
Either throw an error here if this is unspecified, or assign a default and print a warning saying that a default value was used e.g. "world_frame_id param was not specified. Using ROSWorld as the world_frame_id".

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/src/simulate.cpp:
```diff
@@ -45,26 +157,57 @@ int main(int argc, char** argv)
+  n.getParam("/world_frame_id", world_frame_id);
+
+  /// setting target marker color and scale
+  std_msgs::ColorRGBA grey = loadColor(n, "grey");
```
If target marker color is always grey then this does not need to be a parameter and can be hardcoded. 

If you actually want a user to change the colour, then rename this param to target_color. The same also applies to the params "blue" and "orange" which should be named "world_origin_color" and "cam_color" respectively.

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/src/simulate.cpp:
```diff
@@ -45,26 +157,57 @@ int main(int argc, char** argv)
-  marker_position.y = 0;
-  marker_position.z = 0;
+  geometry_msgs::Point starting_target_position = loadPoint(n, "starting_target_position");
+  // geometry_msgs::Quaternion starting_target_orientation = loadOrientation(n, "starting_target_orientation");
```
Remove stale comment.

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/src/simulate.cpp:
```diff
@@ -45,26 +157,57 @@ int main(int argc, char** argv)
+
+  /// making a line of targets and adding to server
+  int num_targets;
+  n.getParam("num_targets_in_line", num_targets);
```
You can shorten this param name to "num_targets"

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/src/simulate.cpp:
```diff
@@ -45,26 +157,57 @@ int main(int argc, char** argv)
+  /// adding camera
+  rviz_simulator::CameraProperties camera_properties = loadCameraProperties(n);
+  geometry_msgs::Point starting_camera_postion = loadPoint(n, "starting_camera_positon");
+  // geometry_msgs::Quaternion starting_camera_orientation = loadOrientation(n, "starting_camera_orientation");
```
Remove this comment if unused.

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/src/camera.cpp:
```diff
@@ -269,18 +281,14 @@ std::vector<Eigen::Vector2d> Camera::processCorners(Eigen::Affine3d camera_extri
 }
 
-Eigen::Vector2d Camera::toPixelCoordinates(Eigen::Affine3d camera_extrinsics, Eigen::Vector4d point)
+Eigen::Vector2d Camera::calculatePixelCoords(Eigen::Affine3d camera_T_target, Eigen::Vector4d point_in_tag)
```
Should rename this to point_in_target for consistency.

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/include/rviz_simulator/camera.h:
```diff
@@ -0,0 +1,231 @@
+
+private:
+  CameraProperties camera_properties_;  /// Camera properties
+  double target_x_length_;              /// The real-world x (horizontal) length of a target
```
Including target length and width in the camera class assumes that all targets will have the same properties but in reality targets can be of differing sizes. Add a TODO here to move this property to the target class in order to make this configurable for each target. You don't need to do this now as it will be a larger overhaul of the current implementation, which works anyway.

chris24sahadeo replied:
TODO noted.

sarika93 commented on rviz_simulator/src/camera.cpp:
```diff
@@ -0,0 +1,497 @@
+  this->camera_properties_ = camera_properties;
+  this->camera_properties_.populateCameraMatrix();
+
+  this->target_x_length_ = target_scale;
```
The purpose of including target_x and target_y is to allow them to be different values if they are not square. Assigning them to the same variable assumes they are square. Use target_scale_x and target_scale_y instead; you'll need an extra param.

chris24sahadeo replied:
TODO added to fix it later on.

sarika93 commented on rviz_simulator/src/camera.cpp:
```diff
@@ -0,0 +1,497 @@
+  if (pixelCoords.x() >= 0 && pixelCoords.x() < this->camera_properties_.image_width && pixelCoords.y() >= 0 &&
+      pixelCoords.y() < this->camera_properties_.image_height)
+  {
+    ROS_INFO_STREAM("Valid corner coordinate\n");
```
This is a misleading message as there are several checks used for determining if a corner is valid. Instead use "Corner is within image" or something similar.

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/src/camera.cpp:
```diff
@@ -0,0 +1,497 @@
+  out << YAML::EndMap;
+}
+
+void Camera::ObjPointsInTargetToYAML(YAML::Emitter& out)
```
This is not formatted according to the Design document, and should be fixed to match that format.

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/src/camera.cpp:
```diff
@@ -0,0 +1,497 @@
+      {
+        detections_out << YAML::BeginMap;  // target
+        detections_out << YAML::Key << "targetID" << YAML::Value << std::to_string(i);
+        detections_out << YAML::Key << "corners";
```
This is not formatted according to the design document. Fix to match the spec we decided on.

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/src/camera.cpp:
```diff
@@ -0,0 +1,497 @@
+      if (!corners.empty())
+      {
+        detections_out << YAML::BeginMap;  // target
+        detections_out << YAML::Key << "targetID" << YAML::Value << std::to_string(i);
```
You are missing the "size" field, which should included (according to the design document).

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/src/camera.cpp:
```diff
@@ -0,0 +1,497 @@
+  out << YAML::BeginMap;  // camera matrix map
+  out << YAML::Key << "rows" << YAML::Value << 3;
+  out << YAML::Key << "cols" << YAML::Value << 3;
+  out << YAML::Key << "data" << YAML::Value;
```
Does not conform to format specified in the design document.

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/src/camera.cpp:
```diff
@@ -0,0 +1,497 @@
+  out << YAML::Key << "rows" << YAML::Value << 1;
+  out << YAML::Key << "cols" << YAML::Value << 5;
+  out << YAML::Key << "data" << YAML::Value;
+  out << YAML::BeginSeq;  // distortion coefficients data sequence
```
Does not conform to format specified in the design document.

chris24sahadeo replied:
Fixed.

sarika93 reviewed (changes requested):
Requested a few changes. When I ran this there was one instance where I was unable to generate an detections even though there was a tag within range of the camera. I was unable to recreate this afterwards though; if I am able to, I will let you know. 

Also you have some very long variable names. Generally, you want names to be clear but with fewer words; abbreaviations are fine once they are clear e.g. dist instead of distance.

chris24sahadeo committed:
Fixing PR changes

chris24sahadeo committed:
Requested changes made.

sarika93 commented on rviz_simulator/launch/simulate.launch:
```diff
@@ -1,13 +1,15 @@
   <!-- loading values into the ros parameter server -->
   <rosparam command="load" file="$(find rviz_simulator)/config/initialize_simulator.yaml" />
 
   <!-- adding a command line arg for camera intrinsics file -->
```
Remove stale comment

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/launch/simulate.launch:
```diff
@@ -1,13 +1,15 @@
-  <arg name="camera_intrinsics_file" default="photoneo_camera_intrinsics.yaml"/>
   <rosparam command="load" file="$(find rviz_simulator)/config/$(arg camera_intrinsics_file)" />
   
   <!-- adding a command line arg for debug mode -->
```
Remove stale comment

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/src/simulate.cpp:
```diff
@@ -155,57 +218,75 @@ int main(int argc, char** argv)
-  std_msgs::ColorRGBA blue = loadColor(n, "blue");
-  double target_scale;
-  n.getParam("target_scale", target_scale);
+  double target_scale = 0.1;
```
Tag size should not be hardcoded. This should be a param as was done previously. However, there should be two params instead of one; one for target_x_scale and one for target_y_scale. You can use a default of target_x_scale = target_y_scale = 0.1.

chris24sahadeo replied:
I made the param in the `initialize_simulator.yaml` file a sequence of length 3 (length, width, depth), example:
`target_size: [0.1, 0.1, 0.01]`
which is read into a `std::vector<double>` from the parameter server in the `simulate.cpp` file (since rosparam cpp can automatically read vectors with no additional code)

The depth is theoretically 0 since the target is a plane, so I gave it a depth of 1cm and set the target class to accept the vector and initialize the visualization marker dimensions accordingly.

sarika93 commented on rviz_simulator/src/camera.cpp:
```diff
@@ -57,13 +57,15 @@ Camera::Camera(const std::string marker_frame_id, const std::string marker_name,
+  // TODO
+  // move target length properties to the target class in order to make this configurable for each target.
   this->target_x_length_ = target_scale;
   this->target_y_length_ = target_scale;
```
x and y should not be assigned the same value as this assumes the target is always square. Your TODO addresses moving these properties into the target class so this still needs to be fixed.

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/src/camera.cpp:
```diff
@@ -309,10 +314,10 @@ bool Camera::isWithinImage(Eigen::Vector2d pixelCoords)
       pixelCoords.y() < this->camera_properties_.image_height)
   {
-    ROS_INFO_STREAM("Valid corner coordinate\n");
+    ROS_INFO_STREAM("Corner is within image\n");
```
ROS_INFO doesn't need "\n" specified

chris24sahadeo replied:
Fixed all instances.

sarika93 commented on rviz_simulator/src/simulate.cpp:
```diff
@@ -49,62 +51,123 @@
+    }
+    else  // invalid point name
+    {
+      ROS_ERROR_STREAM(point_name << " is an invalid param name.");
```
This message is not clear. Instead use:
ROS_ERROR_STREAM("The param " << point_name << "was not found or is of an invalid type. Exiting application!");

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/src/simulate.cpp:
```diff
@@ -49,62 +51,123 @@
+  }
+  else  // point param not specified
+  {
+    ROS_WARN_STREAM(point_name << " param was not specified. Using default values.");
```
Only apply this warning where "point_name" is "starting_target_position" or "starting_camera_position" since default values are only used in those cases.
Currently, in other cases, you will get both this warning and then the error below which will be confusing for a user.

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/src/simulate.cpp:
```diff
@@ -49,62 +51,123 @@
+  std::vector<double> rotation_matrix_col_major;
+  if (!n.getParam(orientation_name, rotation_matrix_col_major))
+  {
+    ROS_WARN_STREAM(orientation_name << " param was not specified. Using default values.");
```
See comment from getting the "point_name" param. That also applies here.

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/src/simulate.cpp:
```diff
@@ -49,62 +51,123 @@
+    }
+    else
+    {
+      ROS_ERROR_STREAM(orientation_name << " is an invalid param name.");
```
See comment from the corresponding logic in getting the "point_name" param. That also applies here.

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/src/simulate.cpp:
```diff
@@ -49,62 +51,123 @@
-  n.getParam("camera_name", camera_properties.camera_name);
+  if (!n.getParam("image_width", camera_properties.image_width))
+  {
+    ROS_ERROR("image_width camera property not specified.");
```
Let the user know that you are exiting the application in this error message.

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/src/simulate.cpp:
```diff
@@ -49,62 +51,123 @@
+
+  if (!n.getParam("image_height", camera_properties.image_height))
+  {
+    ROS_ERROR("image_height camera property not specified.");
```
Let the user know that you are exiting the application in this error message.

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/src/simulate.cpp:
```diff
@@ -49,62 +51,123 @@
+
+  if (!n.getParam("camera_name", camera_properties.camera_name))
+  {
+    ROS_ERROR("camera_name camera property not specified.");
```
Let the user know that you are exiting the application in this error message.

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/src/simulate.cpp:
```diff
@@ -49,62 +51,123 @@
-  n.getParam("camera_matrix/data", camera_matrix);
+  if (!n.getParam("camera_matrix/data", camera_matrix))
+  {
+    ROS_ERROR("camera_matrix camera property not specified.");
```
Let the user know that you are exiting the application in this error message.

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/src/simulate.cpp:
```diff
@@ -49,62 +51,123 @@
+
+  if (!n.getParam("distortion_model", camera_properties.distortion_model))
+  {
+    ROS_ERROR("distortion_model camera property not specified.");
```
Let the user know that you are exiting the application in this error message.

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/src/simulate.cpp:
```diff
@@ -49,62 +51,123 @@
+  std::vector<double> distortion_coefficients;
+  if (!n.getParam("distortion_coefficients/data", distortion_coefficients))
+  {
+    ROS_ERROR("distortion_coefficients camera property not specified.");
```
Let the user know that you are exiting the application in this error message.

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/src/simulate.cpp:
```diff
@@ -119,8 +182,8 @@ rviz_simulator::CameraProperties loadCameraProperties(const ros::NodeHandle& n)
   else
   {
     ROS_ERROR("Unknown camera distortion model specified!\n");
```
Let the user know that you are exiting the application in this error message. Also mention only the "plumb_bob" model is currently supported.

chris24sahadeo replied:
Fixed.

sarika93 commented on rviz_simulator/src/simulate.cpp:
```diff
@@ -155,57 +218,75 @@ int main(int argc, char** argv)
+                    starting_target_position, starting_target_orientation, world_origin_color, target_color);
 
-  /// adding camera
+  /// loading camera properties
```
This comment is redundant since the function name is already "loadCameraProperties". You don't need to remove it, but just pointing that out for future reference.

chris24sahadeo replied:
Noted. I chose to remove it.

chris24sahadeo committed:
Addressed all of Sarika's milestone 1b comments

chris24sahadeo committed:
Commit before implementing boot::filesystem

chris24sahadeo committed:
Switched to boost::filesystem

Commit before converting to iterators

sarika93 commented on rviz_simulator/README.md:
```diff
@@ -35,10 +35,34 @@ and 4 nodes in `debug_mode`:
 > `camera_intrinsics_file:=<camera_intrinsics_file_name.yaml>`  
 
-The `simulator` node creates a new folder `"detections_ROS_timestamp"` in the `rviz_simulator` package folder. A new detections folder with a new timestamp is created each time `roslaunch` is run.
+The `simulator` node creates a new folder with the name `"detections_<ROS_timestamp>"` in the `rviz_simulator` package folder.  
```
The comment you are resolving only said you needed to include an example to the *ROS Timestamp*; not the entire detections file. The file sample clutters the README. 

Additionally, in the below example, it looks like the time-stamped name is associated with the detections YAML file and this is not the case. It's actually used in the folder name. 

Instead, if you want to provide detections examples, create an "examples" folder and upload a few examples of the file format to the repo. If you take this approach (which I would recommend), add in comments so anyone who wants to read this through knows what the fields mean and what the format is (e.g. if matrices are row or column major).

sarika93 commented on rviz_simulator/include/rviz_simulator/camera.h:
```diff
@@ -39,8 +39,9 @@
 // for making directories
-#include <sys/stat.h>
-#include <sys/types.h>
+// #include <sys/stat.h>
```
If these are no longer used, just remove them, rather than commenting it out.

sarika93 commented on rviz_simulator/src/camera.cpp:
```diff
@@ -107,17 +108,17 @@ void Camera::makeOutputDirectories()
   // if detections root directory does not exist then create it
-  struct stat info;
-  if (stat(detections_root_folder.c_str(), &info) != 0)
+  // struct stat info;
```
Remove this commented out line.

sarika93 commented on rviz_simulator/src/camera.cpp:
```diff
@@ -107,17 +108,17 @@ void Camera::makeOutputDirectories()
-  if (mkdir(this->output_folder_path_.c_str(), 0777) == -1)
+  if (!boost::filesystem::create_directories(this->output_folder_path_))
     ROS_ERROR_STREAM("Error :  " << strerror(errno) << std::endl);
 
```
Nit: Remove extra line space for consistency

sarika93 commented on rviz_simulator/src/simulate.cpp:
```diff
@@ -64,22 +64,23 @@ geometry_msgs::Point loadPoint(const ros::NodeHandle& n, const std::string& poin
     else  // invalid point name
     {
-      ROS_ERROR_STREAM(point_name << " is an invalid param name.");
+      ROS_ERROR_STREAM("The param " << point_name << "was not found or is of an invalid type. Exiting application!");
```
Nit: Add in a space before "was" for cleaner formatting.

sarika93 reviewed (commented):
Some minor things to fix. Verifying pixel values before I approve.

sarika93 reviewed (approved):
Approving after testing (which admittedly wasn't super rigorous), with the assumption that the fixes from the previous review will be made before merging.
